pennylane/templates/state_preparations/mottonen.py

```diff
@@ -260,45 +240,34 @@ def MottonenStatePreparation(state_vector, wires):
         "".format(expected_shape, get_shape(state_vector)),
     )
 
-    # check if state_vector is normalized
+    # TODO: delete when tape is new core
     if isinstance(state_vector[0], Variable):
-        state_vector_values = [s.val for s in state_vector]
-        norm = np.sum(np.abs(state_vector_values) ** 2)
-    else:
-        norm = np.sum(np.abs(state_vector) ** 2)
+        state_vector = np.array([s.val for s in state_vector])
+
+    # check if normalized
+    norm = np.sum(np.abs(state_vector) ** 2)
     if not np.isclose(norm, 1.0, atol=1e-3):
         raise ValueError("'state_vector' has to be of length 1.0, got {}".format(norm))
 
     #######################
 
-    # Change ordering of indices, original code was for IBM machines
-    state_vector = np.array(state_vector).reshape([2] * n_wires).T.flatten()[:, np.newaxis]
-    state_vector = sparse.dok_matrix(state_vector)
-
-    a = sparse.dok_matrix(state_vector.shape)
-    omega = sparse.dok_matrix(state_vector.shape)
+    # change ordering of indices, original code was written for IBM machines
```

Comment:
This is pretty minor, but from my past experience, reversing qubit indices can be a recipe for disaster :dizzy_face: . It also leads to the `draw` function producing a circuit that is "upside down" when compared to the original Möttönen et al. figures as well as the [graphic for the template shown in the docs](https://pennylane.readthedocs.io/en/stable/introduction/templates.html#state-preparations). How non-trivial would a rewrite be?

Reply:
I'll give it a try, might not be too bad...

Reply:
I tried for hours to rewrite the code, but just couldn't get it to work. As a compromise, I replaced the reordering of the input state by just reversing the wires, which is a lot faster. You think that's fine?

Reply:
Yes if it's a pain I wouldn't spend more time on it. The circuits come out "right side up" now :)